brayns/common/volume/VolumeHandler.cpp

```diff
@@ -28,60 +28,173 @@
-    if( _memoryMapPtr == MAP_FAILED )
+void VolumeHandler::attachVolumeToFile( const float timestamp, const std::string& volumeFile )
+{
+    _volumeDescriptors[ timestamp ].reset(
+        new VolumeDescriptor( volumeFile, _volumeParameters->getDimensions( )));
+    for( const auto& volumeDescriptor: _volumeDescriptors )
     {
-        _memoryMapPtr = 0;
-        ::close( _cacheFileDescriptor );
-        BRAYNS_THROW( std::runtime_error( "Failed to attach " + volumeFile ));
+        _timestampRange.x() = std::min( _timestampRange.x(), volumeDescriptor.first );
+        _timestampRange.y() = std::max( _timestampRange.y(), volumeDescriptor.first );
     }
+    BRAYNS_INFO << "Attached " << volumeFile << " to timestamp "
+                << timestamp << " " << _timestampRange << std::endl;
+}
 
-    _size = sb.st_size;
+void* VolumeHandler::getData( const float timestamp )
+{
+    const float ts = _getBoundedTimestamp( timestamp );
+    if( ts != _currentTimestamp &&
+        _volumeDescriptors.find( ts ) != _volumeDescriptors.end( ))
+    {
+        if( _volumeDescriptors.find( _currentTimestamp ) != _volumeDescriptors.end( ))
+            _volumeDescriptors[ _currentTimestamp ]->unmap();
+        _currentTimestamp = ts;
+        _volumeDescriptors[ _currentTimestamp ]->map();
+    }
+    return _volumeDescriptors[ _currentTimestamp ]->getMemoryMapPtr();
+}
 
-    BRAYNS_INFO << "Volume size      : " << _size << " bytes ["
-                <<  _size / 1048576 << " MB]" << std::endl;
-    BRAYNS_INFO << "Volume dimensions: " << _dimensions << std::endl;
-    BRAYNS_INFO << "Successfully attached to " << volumeFile << std::endl;
+float VolumeHandler::getEpsilon(
+    const float timestamp ,
+    const Vector3f& elementSpacing,
+    const uint16_t samplesPerRay )
+{
+    const float ts = _getBoundedTimestamp( timestamp );
+    if( _volumeDescriptors.find( ts ) != _volumeDescriptors.end( ))
+    {
+        Vector3f diag = Vector3f( _volumeDescriptors[ ts ]->getDimensions( )) * elementSpacing;
+        return std::max(diag.x(), std::max( diag.y(), diag.z())) / float( samplesPerRay );
+    }
+    else
+    {
+        BRAYNS_ERROR << "No volume is attached to the specified timestamp: "
+                     << timestamp << std::endl;
+        return 0;
+    }
 }
 
-VolumeHandler::~VolumeHandler()
+const Vector3ui VolumeHandler::getDimensions( const float timestamp )
 {
-    if( _memoryMapPtr )
-        ::munmap( (void *)_memoryMapPtr, _size );
+    float ts = _getBoundedTimestamp( timestamp );
+    if( _volumeDescriptors.find( ts ) != _volumeDescriptors.end( ))
+        return _volumeDescriptors[ ts ]->getDimensions();
+    else
+    {
+        BRAYNS_ERROR << "No volume is attached to the specified timestamp: "
+                     << timestamp << std::endl;
+        return Vector3ui( 0.f );
+    }
+}
 
-    if( _cacheFileDescriptor != -1 )
-        ::close( _cacheFileDescriptor );
+uint64_t VolumeHandler::getSize( const float timestamp )
```

Comment:
FYI, the std::map [] operator is not const so I had to replace it by at() to make the method const.